### PR TITLE
[AIRFLOW-393] Add callback for FTP downloads

### PR DESCRIPTION
**Pull request by akosel**
Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW-393) issues and references them in the PR title.

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

Allows a user to pass a callback when downloading an FTP file. An obvious benefit to this is offering a way to log progress as a file downloads. This __heavily__ borrows from @jlowin's [PR].(https://github.com/apache/incubator-airflow/pull/1704/files)

### Tests
- [x] My PR adds the following unit tests:
Adds tests to verify callback is actually called if passed in and also verifies default behavior (passing retrieved data to the write buffer) is preserved in lieu of a custom callback. Also adds tests for some of the helper functions.

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: airflow/contrib/hooks/ftp_hook.py
Comment (skudriashev) [inline airflow/contrib/hooks/ftp_hook.py:205]: Any tests for callback?

File: airflow/contrib/hooks/ftp_hook.py
Comment (skudriashev) [inline airflow/contrib/hooks/ftp_hook.py]: .format removes lazy formatting for logging. I wouldn't change that.

File: airflow/contrib/hooks/ftp_hook.py
Comment (skudriashev) [inline airflow/contrib/hooks/ftp_hook.py:300]: This is redundant.

File: tests/contrib/hooks/test_ftp_hook.py
Comment (skudriashev) [inline tests/contrib/hooks/test_ftp_hook.py]: Shouldn't you mock `size` instead of `get_size`?

Comment (skudriashev) [reply tests/contrib/hooks/test_ftp_hook.py]: According to https://docs.python.org/2/library/ftplib.html#ftplib.FTP.size `size` returns `int`, not `str`.

**Review by skudriashev — CHANGES_REQUESTED** [review]

File: airflow/contrib/hooks/ftp_hook.py
Comment (akosel) [reply airflow/contrib/hooks/ftp_hook.py:205]: Added

**Comment by akosel** [comment]
Better late than never...I'm not sure if @skudriashev is still a maintainer of Airflow, but if not, could I get another review/stamp?

File: airflow/contrib/hooks/ftp_hook.py
Comment (ashb) [inline airflow/contrib/hooks/ftp_hook.py:174]: Can we have some docs on what the callback is called with (what args), when it is called, what doesn't happen normally if you use a callback (i.e. I think it won't write any files.)

Also possibly as implemeneted now it would be better to make callback at local_full_path_or_bucket to be mutually exclusive - if a callback is passed the local_* argument is ignored.

**Review by ashb — CHANGES_REQUESTED** [review]

File: airflow/contrib/hooks/ftp_hook.py
Comment (akosel) [reply airflow/contrib/hooks/ftp_hook.py:174]: Good idea! I also added an example section to the docstring. If that seems excessive, I can remove it. Emphasis on the need to handle writing to the file inside of the custom callback if it is used.

**Comment by Fokko** [comment]
LGTM. Thanks @akosel

**Pull request merged**
